### PR TITLE
Adjust concurrency

### PR DESCRIPTION
**Pull request by mellowdrifter**
Few things:
1 - If number of files < 4, no need to spin up a default of 4. Just set threads to number of files
2 - Only spin up a worker when there is work to do. i.e. do not create a worker accepting a channel of files, just dispatch a worker for each file
3 - The amount of workers in #2 is controlled by the semaphore, so no need for a separate mechanism to control this
4 - Make the displaying of records to stdout a Boolean flag.

All the above allows us to use a wait group instead of a weird loop :)

### CODE REVIEW

**Comment by mellowdrifter** [comment]
Move away from a receive channel. Tested and this is much faster with multiple files. addRecord method is an atomic method allowing multiple goroutines to add to the slice

File: rpsl-tools/main.go [hunk 1]
```diff
@@ -71,11 +71,18 @@ func getReader(fn string) (io.Reader, error) {
...
+	return ch
+}
+
 // parseFile reads and parses files as their filenames arrive on
 // the input channel. A parse error will abort processing for the
 // corresponding file and move to the next one
-func parseFile(ic <-chan string, rc chan<- *rppb.Record, ec chan<- bool) {
+func parseFile(ic <-chan string, ch chan *rppb.Record) {
 	for fn := range ic {
+		//files++
```
Comment (morrowc) [inline rpsl-tools/main.go]: probably removable comment line.

File: rpsl-tools/main.go [hunk 2]
```diff
@@ -130,47 +142,27 @@ func main() {
...
+	}
+	fmt.Println("about to wait")
+	fmt.Println("finished waiting")
+
+	for _, v := range allRecords {
+		displayRecords(v)
 	}
 
-	close(rc)
+	//fmt.Printf("Processed with %d records", len(rc))
```
Comment (morrowc) [inline rpsl-tools/main.go]: probably also removable.

File: rpsl-tools/main.go [hunk 3]
```diff
@@ -130,47 +142,27 @@ func main() {
...
-			}
-		default:
-		}
+	// Start the parsing/worker thread
+	// TODO: if len(files) < threads, just use len(files)
+	for i := 0; i < *threads; i++ {
+		records := prepWorker(ic)
+		allRecords = append(allRecords, records)
+	}
+	fmt.Println("about to wait")
```
Comment (morrowc) [inline rpsl-tools/main.go]: might not need these.

**Review by morrowc — CHANGES_REQUESTED** [review]

File: rpsl-parser/proto/rpsl.pb.go [hunk 4]
```diff
@@ -26,7 +18,7 @@ var _ = math.Inf
 // is compatible with the proto package it is being compiled against.
 // A compilation error at this line likely means your copy of the
 // proto package needs to be updated.
-const _ = proto.ProtoPackageIsVersion2 // please upgrade the proto package
+const _ = proto.ProtoPackageIsVersion3 // please upgrade the proto package
```
Comment (morrowc) [inline rpsl-parser/proto/rpsl.pb.go:21]: why did the generated pb.go change?
it looks like maybe the .proto file changed from proto2 -> proto3 requirements?

**Review by morrowc — CHANGES_REQUESTED** [review]

File: rpsl-tools/main.go (hunk 1 above)
Comment (mellowdrifter) [reply rpsl-tools/main.go]: resync :) I already removed these and these are based on outdated files

File: rpsl-parser/proto/rpsl.pb.go (hunk 4 above)
Comment (mellowdrifter) [reply rpsl-parser/proto/rpsl.pb.go:21]: no that didn't change. I merely reran protoc and the generator itself is a later version

File: rpsl-tools/main.go [hunk 5]
```diff
@@ -74,53 +76,59 @@ func getReader(fn string) (io.Reader, error) {
...
+	rpsl.Parse(rdr, rc)
+	<-sem
 }
 
-// Verify that there files requested exist, open each in a goroutine and feed
+// Verify that the files requested exist, open each in a goroutine and feed
 // those to the rpsl-parser library, returning each record to a channel for
 // disposition in the final data structure to be loaded into a DB.
 func main() {
+	//defer profile.Start(profile.CPUProfile, profile.ProfilePath(".")).Stop()
```
Comment (morrowc) [inline rpsl-tools/main.go:129]: these probably can go, right?

**Review by morrowc — APPROVED** [review]

**Pull request merged**